component/config.go

```diff
@@ -31,46 +33,131 @@ type ConfigValidator interface {
+			} else {
+				err = errors.Join(err, suberr.err)
+			}
+		}
+	}
+
+	return err
+}
+
+type pathError struct {
+	err  error
+	path string
 }
 
-func validate(v reflect.Value) error {
+func validate(v reflect.Value) []pathError {
+	errs := []pathError{}
 	// Validate the value itself.
 	switch v.Kind() {
 	case reflect.Invalid:
 		return nil
 	case reflect.Ptr, reflect.Interface:
 		return validate(v.Elem())
 	case reflect.Struct:
-		var errs error
-		errs = multierr.Append(errs, callValidateIfPossible(v))
+		errs = append(errs, pathError{err: callValidateIfPossible(v), path: ""})
 		// Reflect on the pointed data and check each of its fields.
 		for i := 0; i < v.NumField(); i++ {
 			if !v.Type().Field(i).IsExported() {
 				continue
 			}
-			errs = multierr.Append(errs, validate(v.Field(i)))
+			subpathErrs := validate(v.Field(i))
+			for _, err := range subpathErrs {
+				field := v.Type().Field(i)
+				var fieldName string
+				if tag, ok := field.Tag.Lookup("mapstructure"); ok {
+					tags := strings.Split(tag, ",")
+					if len(tags) > 0 {
+						fieldName = tags[0]
+					}
+				}
+				// Even if the mapstructure tag exists, the field name may not
+				// be available, so set it if it is still blank.
+				if len(fieldName) == 0 {
+					fieldName = strings.ToLower(field.Name)
+				}
+
+				var path string
+				if len(err.path) > 0 {
+					path = strings.Join([]string{fieldName, err.path}, "::")
+				} else {
+					path = fieldName
+				}
+				errs = append(errs, pathError{
+					err:  err.err,
+					path: path,
+				})
+			}
 		}
 		return errs
 	case reflect.Slice, reflect.Array:
-		var errs error
-		errs = multierr.Append(errs, callValidateIfPossible(v))
+		errs = append(errs, pathError{err: callValidateIfPossible(v), path: ""})
 		// Reflect on the pointed data and check each of its fields.
 		for i := 0; i < v.Len(); i++ {
-			errs = multierr.Append(errs, validate(v.Index(i)))
+			subPathErrs := validate(v.Index(i))
+
+			for _, err := range subPathErrs {
+				var path string
+				if len(err.path) > 0 {
+					path = strings.Join([]string{strconv.Itoa(i), err.path}, "::")
+				} else {
+					path = strconv.Itoa(i)
+				}
+
+				errs = append(errs, pathError{
+					err:  err.err,
+					path: path,
+				})
+			}
 		}
 		return errs
 	case reflect.Map:
-		var errs error
-		errs = multierr.Append(errs, callValidateIfPossible(v))
+		errs = append(errs, pathError{err: callValidateIfPossible(v), path: ""})
 		iter := v.MapRange()
 		for iter.Next() {
-			errs = multierr.Append(errs, validate(iter.Key()))
-			errs = multierr.Append(errs, validate(iter.Value()))
+			keyErrs := validate(iter.Key())
+			valueErrs := validate(iter.Value())
+			var key string
+
+			if str, ok := iter.Key().Interface().(string); ok {
+				key = str
+			} else if stringer, ok := iter.Key().Interface().(fmt.Stringer); ok {
+				key = stringer.String()
+			} else {
+				key = fmt.Sprintf("[%T key]", iter.Key().Interface())
```

Comment:
I'm not sure what we want to put here, but for keys that can't be made into a reasonable string representation, I've just chosen to print the type so we have something displayed. There are tests that show how this looks.

Reply:
We could also use `%v` or `%#v`

Reply:
I avoided printing the value because it produces odd-looking error messages like:

```
mapkeychild::[component.configChildStruct{Child:component.errConfig{err:(*errors.errorString)(0xc000039c80)}, ChildPtr:(*component.errConfig)(nil)}]::child: child key error
```

We can probably be smarter about this by doing additional reflection to determine whether we're looking at a primitive or complex type, my initial goal was just to keep it simple for this iteration. I'll play with this.

Reply:
I arrived at something I think is simple enough, take a look in the latest commit.

Reply:
Looks reasonable!